config/src/keys.rs

```diff
@@ -99,22 +45,66 @@ where
 impl<T> KeyPair<T>
 where
     T: PrivateKey + Serialize + ValidKeyStringExt,
-    T::PublicKeyMaterial: DeserializeOwned + 'static + Serialize + ValidKeyStringExt,
 {
+    /// This transforms a private key into a keypair data structure.
     pub fn load(private_key: T) -> Self {
         let public_key = private_key.public_key();
         Self {
-            private_key: PrivateKeyContainer::Present(private_key),
+            private_key: Some(private_key),
             public_key,
         }
     }
 
-    pub fn public(&self) -> &T::PublicKeyMaterial {
-        &self.public_key
-    }
-
-    /// Beware, this destroys the private key from this NodeConfig
+    /// Takes the key from the data structure, calling this function a second time will return None.
     pub fn take_private(&mut self) -> Option<T> {
         self.private_key.take()
     }
+
+    /// Returns the public key part. This always work, even after the private key was taken.
+    pub fn public_key(&self) -> T::PublicKeyMaterial {
+        self.public_key.clone()
+    }
+}
+
+//
+// Serialization and Deserialization functions
+//
+
+/// Serialization for a KeyPair only serializes the private key part (if present).
+impl<T> Serialize for KeyPair<T>
+where
+    T: PrivateKey + Serialize + ValidKeyStringExt,
+{
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+        T: Serialize + ValidKeyStringExt,
+    {
+        match &self.private_key {
+            Some(key) => key.serialize(serializer),
+            None => serializer.serialize_str(""),
```

Comment:
I would argue this should probably be an error.... and if you agree, then we could do skip serialize in serde.

Reply:
I'm not sure what are all the use cases for this, for example in the future maybe we'll want to allow loading the private key from either the config (for quick testing) or from an env variable or from some other mechanism